### PR TITLE
feat(modellib): deduplicate file download [POFSP-1137]

### PR DESCRIPTION
**Pull request by polomani**
After the recent change for multi-file model support https://github.com/cognitedata/dotnet-simulator-utils/pull/282 we need to add some logic for performance optimisation.
The model files (and their dependencies) need to be deduplicated on download. That means if there is the same file locally already, we don't have to download it again, and only link it to the existing `Path`.
To achieve that we compare the list of files in the state store to the files needed for a given model.

Additional change: include `.Downloaded` check for the model deps, in addition to the main file.

### CODE REVIEW

File: Cognite.Simulator.Utils/ModelLibraryBase.cs [hunk 1]
```diff
@@ -564,6 +564,30 @@ private static void CreateDirectoryIfNotExists(string directoryPath)
...
+                .Where(s => !string.IsNullOrEmpty(s.Value.FilePath) && filesPathsInSubfolders.ContainsKey(s.Value.FilePath))
+                .ToDictionarySafe(s => s.Value.CdfId, s => s.Value.FilePath);
+
+            var dependencyFiles = _state
+                .SelectMany(s => s.Value.DependencyFiles)
+                .Where(f => !string.IsNullOrEmpty(f.FilePath) && filesPathsInSubfolders.ContainsKey(f.FilePath))
+                .ToDictionarySafe(f => f.Id, f => f.FilePath);
+
+            return mainModelFiles.Union(dependencyFiles)
+                .ToDictionary(kvp => kvp.Key, kvp => kvp.Value);
```
[the remark this answers is not in the document]
Comment (polomani) [reply Cognite.Simulator.Utils/ModelLibraryBase.cs:588]: it's forbidden for a file to have itself as a dependency on the API level, and it's generally not a use case

[the remark this answers is not in the document]
Comment (polomani) [reply Cognite.Simulator.Utils/ModelLibraryBase.cs:588]: answered on a different thread already. no cyclic deps allowed

**Review by abdullah-cognite — COMMENTED** [review]
Looks good overall, I got a bit lost in the integration tests 😅 though, would prefer a unit test for `GetLocalFilesCache`

File: Cognite.Simulator.Utils/ModelLibraryBase.cs [hunk 2]
```diff
@@ -579,48 +603,54 @@ private async Task<bool> DownloadAllModelFilesAsync(T modelState)
             }
             modelState.DownloadAttempts++;
 
-            var fileIds = modelState.GetPendingDownloadFileIds();
+            var (idsToDownload, existingLocalFiles) = modelState.DeduplicateDownloadFileIds(GetLocalFilesCache());
```
[the remark this answers is not in the document]
Comment (polomani) [reply Cognite.Simulator.Utils/ModelLibraryBase.cs:606]: I agree, but this is better to handle as a separate ticket. https://cognitedata.atlassian.net/jira/software/c/projects/POFSP/boards/852/backlog?selectedIssue=POFSP-1159

File: Cognite.Simulator.Utils/ModelLibraryBase.cs [hunk 3]
```diff
@@ -564,6 +564,30 @@ private static void CreateDirectoryIfNotExists(string directoryPath)
...
         }
 
+        /// <summary>
+        /// Returns a map of File IDs to local file paths for all model files stored locally.
+        /// This includes main model files and dependency files.
+        /// </summary>
+        private Dictionary<long, string> GetLocalFilesCache()
+        {
+            // get all files in the model folder and its subfolders (but not in subfolders of subfolders)
+            var filesPathsInSubfolders = Directory.EnumerateDirectories(_modelFolder)
```
[the remark this answers is not in the document]
Comment (polomani) [reply Cognite.Simulator.Utils/ModelLibraryBase.cs:574]: all of them seem to be very much critical and I don't see how we would handle them gracefully (only to rethrow the same error, but there is no much grace in it :)

File: Cognite.Simulator.Tests/UtilsTests/ModelLibraryExternalDepsTest.cs [hunk 4]
```diff
@@ -262,5 +263,73 @@ public async Task TestModelLibraryWithExternalDependenciesLoadFailure()
...
+                })
+                .ToArray();
+            Assert.Equivalent(expectedDependencyFiles, v1.DependencyFiles);
+            Assert.Equivalent(expectedDependencyFiles, v2.DependencyFiles);
+
+            foreach (var mocker in endpointMockTemplates)
+            {
+                mocker.AssertCallCount();
+            }
+
```
[the remark this answers is not in the document]
Comment (polomani) [reply Cognite.Simulator.Tests/UtilsTests/ModelLibraryExternalDepsTest.cs:325]: ok, will extend this to cover that situation

File: Cognite.Simulator.Utils/StateUtils.cs [hunk 5]
```diff
@@ -97,5 +97,38 @@ public static void DeleteLocalDirectory(string path)
...
+
+            var dependencyFiles = state
+                .SelectMany(s => s.Value.DependencyFiles)
+                .Where(f => !string.IsNullOrEmpty(f.FilePath) && localFilePaths.Contains(f.FilePath))
+                .ToDictionarySafe(f => f.Id, f => f.FilePath);
+
+            return mainModelFiles.Union(dependencyFiles)
+                .ToDictionary(kvp => kvp.Key, kvp => kvp.Value);
+        }
     }
```
[the remark this answers is not in the document]
Comment (polomani) [reply Cognite.Simulator.Utils/StateUtils.cs:133]: answered in the private message, this logic is not correct as it puts empty values in the dict.
also, it doesn't reduce the number of iterations (it's still O(N)) it just makes it less obvious

File: Cognite.Simulator.Utils/ModelLibraryBase.cs [hunk 6]
```diff
@@ -564,6 +564,30 @@ private static void CreateDirectoryIfNotExists(string directoryPath)
             }
         }
 
+        /// <summary>
+        /// Returns a map of File IDs to local file paths for all model files stored locally.
+        /// This includes main model files and dependency files.
+        /// </summary>
+        private Dictionary<long, string> GetLocalFilesCache()
```
[the remark this answers is not in the document]
Comment (abdullah-cognite) [reply Cognite.Simulator.Utils/ModelLibraryBase.cs]: The reason I asked for this was that I find the request mocker in the integration test a bit hard to follow at times. So I just wanted unit tests for the edge cases like below that just focusses on this function. But if you think this is over engineering then feel free to ignore, I dont have any strong opinion on including this. 

````
public void BuildFilesCache_DependencyFilesOnly_ReturnsCorrectMapping()
    {
        // Arrange
        var localFiles = new[] { "models/deps/dep1.out", "models/deps/dep2.out" };
        
        _testState.Add(new FileStatePoco 
        { 
            Value = new FileStatePoco 
            { 
                CdfId = 123L, 
                FilePath = "models/nonexistent/main.out", // Hopefully this won't be included
                DependencyFiles = new List<DependencyFile>
                {
                    new DependencyFile { Id = 789L, FilePath = "models/deps/dep1.out" },
                    new DependencyFile { Id = 101L, FilePath = "models/deps/dep2.out" }
                }
            } 
        });

        // Act
        var result = _service.BuildFilesCache(localFiles, _testState);

        // Assert
        Assert.That(result.Count, Is.EqualTo(2));
        Assert.That(result[789L], Is.EqualTo("models/deps/dep1.out"));
        Assert.That(result[101L], Is.EqualTo("models/deps/dep2.out"));
    }

**Review by abdullah-cognite — APPROVED** [review]

File: Cognite.Simulator.Utils/ModelLibraryBase.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (polomani) [reply Cognite.Simulator.Utils/ModelLibraryBase.cs]: That is a good point! it's is, however, a broader issue (potentially), so I made a separate ticket for it: https://cognitedata.atlassian.net/jira/software/c/projects/POFSP/boards/852/backlog?selectedIssue=POFSP-1166

File: Cognite.Simulator.Utils/ModelLibraryBase.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (polomani) [reply Cognite.Simulator.Utils/ModelLibraryBase.cs]: I did the change, it is not exactly how you wrote it, but does similar thing
Maybe you are right.
Need to move forward with this, so I am closing this

**Review by einarmo — COMMENTED** [review]
:unicorn: - Looks good. I have one nit, but it isn't a blocker.
